processor/metricstransformprocessor/metrics_transform_processor.go

```diff
@@ -60,29 +119,26 @@ func (mtp *metricsTransformProcessor) ProcessMetrics(_ context.Context, md pdata
 
 	for i := range mds {
 		data := &mds[i]
-		nameToMetricMapping := make(map[string]*metricspb.Metric, len(data.Metrics))
-		for _, metric := range data.Metrics {
-			nameToMetricMapping[metric.MetricDescriptor.Name] = metric
-		}
 
+		nameToMetricMapping := newMetricNameMapping(data)
 		for _, transform := range mtp.transforms {
-			metric, ok := nameToMetricMapping[transform.MetricName]
-			if !ok {
-				continue
-			}
+			matchedMetrics := transform.MetricIncludeFilter.getMatches(nameToMetricMapping)
+			for _, metric := range matchedMetrics {
+				metricName := metric.MetricDescriptor.Name
 
-			if transform.Action == Insert {
-				metric = proto.Clone(metric).(*metricspb.Metric)
-				data.Metrics = append(data.Metrics, metric)
-			}
+				if transform.Action == Insert {
+					metric = proto.Clone(metric).(*metricspb.Metric)
+					data.Metrics = append(data.Metrics, metric)
+				}
 
-			mtp.update(metric, transform)
+				mtp.update(metric, transform)
 
-			if transform.NewName != "" {
-				if transform.Action == Update {
-					delete(nameToMetricMapping, transform.MetricName)
+				if transform.NewName != "" {
+					if transform.Action == Update {
+						nameToMetricMapping.remove(metricName, metric)
+					}
+					nameToMetricMapping.add(metric.MetricDescriptor.Name, metric)
```

Comment:
Shouldn't this be `transform.NewName`?

Reply:
The metric name has been transformed at this point so either would work.

I wanted to use `metric.MetricDescriptor.Name` here so this works even if `transform.NewName` supported substitution, which I've now added to this PR - see comment below.